File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def search(node, depth):
            nonlocal result
            if not node: return
            if node.left: search(node.left, depth + 1)
            if node.right: search(node.right, depth + 1)
            if depth > result: result = depth
            return
        
        result = 0
        search(root, 1)
        return result

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result = 0;

    public void search(TreeNode node, int depth) {
        if (node == null) return;
        if (node.left != null) {
            search(node.left, depth + 1);
        }
        if (node.right != null) {
            search(node.right, depth + 1);
        }
        if (result < depth) {
            result = depth;
        }
        return;
    }

    public int maxDepth(TreeNode root) {
        search(root, 1);
        return result;
    }
}